### DISCUSSION TITLE
Possibility to add callback

### DISCUSSION THREAD

**Discussion by AdamMusa** [fastapi/sqlmodel#917] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from sqlmodel import Field, SQLModel


class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None

    def save(self,*args, **kwargs)):
        .....
        return super(Hero,self).save(*args, **kwargs)
```

### Description

It's gonna override or deal like callback the save , update or delete method

### Operating System

macOS

### SQLModel Version

from model.hero import Hero

### Python Version

3.13.0

### Additional Context

Hi, I want to know is that possible to add  a callback before or after save, update,delete in my model?
It would be very cool if someone could answer and given an example.

**Comment by joachimhuet** [comment]
Hey, what do you mean by a 'callback' ? You can pass a function as an argument and call it if that's what you want, if you can be a bit more precise on what you are trying to achieve 😄 

And just a small consideration, but I would suggest you to have your models only for modelling and than have your operations elsewhere. You can have a look at my example project if interested [https://github.com/joachimhuet/sqlmodel-project-example](https://github.com/joachimhuet/sqlmodel-project-example)

**Comment by itsbekas** [comment, marked as answer]
Hey! I think you're looking for [SQLAlchemy's ORM Events](https://docs.sqlalchemy.org/en/20/orm/events.html), namely [Mapper Events](https://docs.sqlalchemy.org/en/20/orm/events.html#mapper-events). You'll likely find a lot more information if you search for information related to SQLAlchemy rather than SQLModel (which uses SQLAlchemy internally), but here's a small example showcasing how these events work:

```py
from sqlmodel import Field, SQLModel, create_engine, Session
from sqlalchemy import event

class Hero(SQLModel, table=True):
    name: str = Field(primary_key=True)
    secret_name: str = Field()

# Option 1: Setting the event listener on the class
def insert_is_done(mapper, connection, target):
    print("I just created a new Hero!")

event.listen(Hero, "after_insert", insert_is_done)

# Option 2: Setting the event listener on the method
@event.listens_for(Hero, "before_insert")
def before_insert(mapper, connection, target):
    print("I'm about to create a new Hero!")

# Create the database
engine = create_engine("sqlite:///example.db")
SQLModel.metadata.create_all(engine)

# Create a Hero
hero_1 = Hero(name="Spider-Man", secret_name="Peter Parker")

with Session(engine) as session:
    session.add(hero_1)
    session.commit()
```

Note that these operations are ran during the `session.commit` operation, not `session.add` (which may or may not be useful for your case). Other operations and a lot more information is available in the SQLAlchemy documentation, which I linked at the beginning of the post.

**Reply by AdamMusa** [reply]
Thank so much @itsbekas

**Reply by itsbekas** [reply]
No Problem! I would just appreciate it if you could mark the question as answered as it might help other users with the same issue and will also let other helpers know this issue is solved 🙂
